includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,313 @@
+
+/**
+ * Class AMP_DOM_Document.
+ *
+ * @since 1.5
+ *
+ * @property DOMXpath $xpath XPath query object for this document.
+ *
+ * Abstract away some of the difficulties of working with PHP's DOMDocument.
+ */
+final class AMP_DOM_Document extends DOMDocument {
+
+	/**
+	 * AMP requires the HTML markup to be encoded in UTF-8.
+	 */
+	const AMP_ENCODING = 'utf-8';
+
+	/**
+	 * Encoding identifier to use for an unknown encoding.
+	 */
+	const UNKNOWN_ENCODING = 'auto';
+
+	/**
+	 * Regular expression pattern to match the http-equiv meta tag.
+	 */
+	const HTTP_EQUIV_META_TAG_PATTERN = '/<meta [^>]*?\s*http-equiv=[^>]*?>[^<]*(?:<\/meta>)?/i';
+
+	/**
+	 * Regular expression pattern to match the charset meta tag.
+	 */
+	const CHARSET_META_TAG_PATTERN = '/<meta [^>]*?\s*charset=[^>]*?>[^<]*(?:<\/meta>)?/i';
+
+	/**
+	 * ID of the hacky charset we need to add to make loadHTML() behave.
+	 */
+	const CHARSET_HACK_ID = '--amp-dom-document-charset--';
+
+	/**
+	 * The original encoding of how the AMP_DOM_Document was created.
+	 *
+	 * This is stored to do an automatic conversion to UTF-8, which is
+	 * a requirement for AMP.
+	 *
+	 * @var string
+	 */
+	private $original_encoding;
+
+	/**
+	 * Associative array of encoding mappings.
+	 *
+	 * Translates HTML charsets into encodings PHP can understand.
+	 *
+	 * @var string[]
+	 */
+	private $encoding_map = [
+		'latin-1' => 'iso-8859-1',
+	];
+
+	/**
+	 * Creates a new AMP_DOM_Document object
+	 *
+	 * @link  https://php.net/manual/domdocument.construct.php
+	 *
+	 * @param string $version  Optional. The version number of the document as part of the XML declaration.
+	 * @param string $encoding Optional. The encoding of the document as part of the XML declaration.
+	 */
+	public function __construct( $version = '', $encoding = null ) {
+		$this->original_encoding = (string) $encoding ?: self::UNKNOWN_ENCODING;
+		parent::__construct( $version ?: '1.0', self::AMP_ENCODING );
+	}
+
+	/**
+	 * Load HTML from a string.
+	 *
+	 * @link  https://php.net/manual/domdocument.loadhtml.php
+	 *
+	 * @param string     $source  The HTML string.
+	 * @param int|string $options Optional. Specify additional Libxml parameters.
+	 * @return bool true on success or false on failure.
+	 */
+	public function loadHTML( $source, $options = 0 ) {
+		$source = $this->maybe_add_head_or_body( $source );
+
+		$this->original_encoding = $this->detect_and_strip_encoding( $source );
+
+		if ( self::AMP_ENCODING !== strtolower( $this->original_encoding ) ) {
+			$source = $this->adapt_encoding( $source );
+		}
+
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$source = str_replace(
+			'<head>',
+			'<head><meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">',
+			$source
+		);
+
+		$success = parent::loadHTML( $source, $options );
+
+		if ( $success ) {
+			$this->encoding = self::AMP_ENCODING;
```

Comment:
Wouldn't `DOMDocument` populate this? Why is this needed?

Reply:
It's one of the many bits I found online that might or might not improve DOMDocument behavior with UTF-8.

However, I don't think this does much, so I'll remove it.